### PR TITLE
feat(react): add `autoAlign` feature to popover

### LINKED ISSUE #11385: Adjust Popover to be visible on first render
Ideal: 

Popover moves into view if original positioning is outside of viewport when the Popover is opened.

### PR DESCRIPTION
**Pull request by abbeyhrt**
Closes #11385 

Adds `autoAlign` prop and corresponding feature to Popover component.  Popover with `autoAlign={true}` will only work on first render so resize and scroll events will currently have no impact on the positioning. 

#### Testing / Reviewing

Review code to make sure the approach is okay and go to the 'Automate' story for Popover to see if it is visible in all four instances

### CODE REVIEW

File: packages/react/src/components/Popover/index.js
Comment (jnm2377) [inline packages/react/src/components/Popover/index.js:81]: just wanna make sure I'm understanding correctly. this is comparing the popover measurements to the page margins? (-16?)

Comment (jnm2377) [reply packages/react/src/components/Popover/index.js:81]: wondering if you could maybe do a one or two liner comment summarizing what `conditions` is for, for maintainers?

**Comment by jnm2377** [comment]
This looks great! 🔥  Left some comments about maybe adding a few more contextual comments for maintainers, but it's not a dealbreaker. The story seems to be working great. I'm wondering if we should add some tests around this functionality, though?

File: packages/react/src/components/Popover/index.js
Comment (abbeyhrt) [reply packages/react/src/components/Popover/index.js:81]: Exactly! The 16 is to account for popover's internal padding. I added a comment to explain, let me know if that helps :)

**Comment by abbeyhrt** [comment]
@jnm2377 Thanks for looking through it! I think the only way to test it is through a VRT test.  Since it's just a visual change, there's no way really to meaningfully test this with RTL. There aren't any VRT tests for Popover yet and I asked @joshblack and he said to hold off on adding one for this variant until it does

**Review by jnm2377 — APPROVED** [review]

File: packages/react/src/components/Popover/next/Popover.stories.js
Comment (joshblack) [inline packages/react/src/components/Popover/next/Popover.stories.js]:
```suggestion
export const AutoAlign = () => {
```

Would this kind of story title make sense? `Automate` wasn't immediately clear to me but let me know if you feel differently!

File: packages/react/src/components/Popover/index.js
Comment (joshblack) [inline packages/react/src/components/Popover/index.js]: Super small nitpick, the convention I think I've been using is:

```suggestion
  forwardRef
) {
  const prefix = usePrefix();
  const { floating } = React.useContext(PopoverContext);
  const ref = useMergedRefs([floating, forwardRef]);
```

File: packages/react/src/components/Popover/index.js
Comment (joshblack) [inline packages/react/src/components/Popover/index.js]: Since this is a context provider, it might help to either `useMemo` the object here to prevent unnecessary child re-renders or pass in `floating` directly since it will have a stable object identity (and prevent re-renders)

File: packages/react/src/components/Popover/index.js
Comment (joshblack) [inline packages/react/src/components/Popover/index.js:25]: Did we ever document the different strategies we were looking at for auto alignment? e.g. `align="auto"` was another approach and I wasn't sure what the process was for `autoAlign`.

Are you hoping to ship this in stable next release or were you hoping to guard this behind a flag / indicate it's unstable?

File: packages/react/src/components/Popover/index.js
Comment (joshblack) [inline packages/react/src/components/Popover/index.js:74]: Will the popover always have this internal padding? I had thought this was specific to the story and wanted to check if that padding changed if this still would work

File: packages/react/src/components/Popover/index.js
Comment (joshblack) [inline packages/react/src/components/Popover/index.js:165]: Will `alignment` always be defined due to L156 above? Or is there a case where it's still `null`?

**Review by joshblack — CHANGES_REQUESTED** [review]
Looking great! Left some comments/questions

File: packages/react/src/components/Popover/next/Popover.stories.js
Comment (abbeyhrt) [reply packages/react/src/components/Popover/next/Popover.stories.js]: Agreed! I like AutoAlign as a title much better

File: packages/react/src/components/Popover/index.js
Comment (abbeyhrt) [reply packages/react/src/components/Popover/index.js:74]: Ugh, dang I hadn't noticed that. I think i would still want to keep it though. The way that rect.x and rect.y work, it will be negative if it goes passed the padding of the page. I wanted to account for that a little bit so I figured 1rem was sensible.

File: packages/react/src/components/Popover/index.js
Comment (abbeyhrt) [reply packages/react/src/components/Popover/index.js:25]: I did not document the different potential strategies. I can document what I ended up choosing  somewhere if you think that's necessary!  

 With `autoAlign` being off by default, I thought it would be okay to include it in the next release without a guard since people will have to opt in to it. I would prefer to not make it officially experimental but I'm down to put a note in the prop description for it

**Review by laurenmrice — APPROVED** [review]
Looks great Abbey! 🎉

**Review by tay1orjones — APPROVED** [review]

**Review by joshblack — APPROVED** [review]
🔥

**Pull request merged**
